File: python/nedb/nesql.py
# SPDX-FileCopyrightText: 2026 INTERCHAINED LLC
# SPDX-License-Identifier: BUSL-1.1
# NEDB · © 2026 INTERCHAINED LLC × Eth-Interchained × Vex (Claude Opus 5)

"""neSQL dialect routing for the Python reference engine.

NEDB speaks NQL **and** SQL. This module decides which one a statement is
written in, so every Python-side door can accept either — the HTTP `/query`
endpoint, `/subscribe`, and any caller holding a `NEDB` instance.

# This is a MIRROR, and mirrors drift

The authoritative router is `rust/nedb-v2/src/nesql.rs`. This file exists
because the Python reference engine cannot call it, not because a second
opinion is wanted — and a second copy of "which language is this" is exactly
the kind of duplication that has silently diverged in this codebase before
(required-vs-optional GROUP BY, the ordering-comparison-against-missing-field
split).

So the vocabularies are asserted identical by
`tests/test_nesql_parity.py`, which reads the Rust source and compares the
keyword lists token for token. If someone adds a statement head to one side,
that test fails rather than the two engines quietly disagreeing about what a
statement MEANS — which is worse than disagreeing about a result, because
nothing looks broken when it happens.

# Routing is structural, not a guess

    NQL  begins with FROM
    SQL  begins with SELECT INSERT UPDATE DELETE EXPLAIN WITH SHOW SET
                     VALUES TABLE BEGIN COMMIT ROLLBACK

PostgreSQL has no statement form that begins with `FROM`, so the leading
keyword PARTITIONS the two vocabularies rather than hinting at them. A first
word in neither is refused NAMING BOTH — never handed to whichever parser
seems likelier, because "seems likelier" is the guess this rule forbids.
"""

from __future__ import annotations

#: Kept in the same order as the Rust lists so a diff between the two is
#: readable by eye as well as by the parity test.
NQL_HEADS = ("FROM",)
SQL_HEADS = (
    "SELECT", "INSERT", "UPDATE", "DELETE", "EXPLAIN", "WITH", "SHOW", "SET",
    "VALUES", "TABLE", "BEGIN", "COMMIT", "ROLLBACK",
)


class DialectError(ValueError):
    """A statement that begins in neither half of neSQL."""


def first_word(statement: str) -> str | None:
    """The leading keyword, uppercased, or None for an empty statement."""
    parts = statement.split()
    if not parts:
        return None
    # A statement may open with a parenthesis — `(SELECT …) UNION …` — and may
    # carry a trailing semicolon on the first token when it is a bare verb.
    word = parts[0].lstrip("(").rstrip(";").upper()
    return word or None


def route(statement: str) -> str:
    """``"nql"`` or ``"sql"``, or raise :class:`DialectError`.

    Returns the same strings `Dialect::name()` returns on the Rust side, so a
    response body built here and one built there are indistinguishable to a
    client.
    """
    head = first_word(statement)
    if head is None:
        raise DialectError("the statement is empty")
    if head in NQL_HEADS:
        return "nql"
    if head in SQL_HEADS:
        return "sql"
    raise DialectError(
        '"%s" does not begin a neSQL statement.\n'
        "neSQL is PostgreSQL's SQL plus NEDB's own clauses, so a statement "
        "starts in one of these two vocabularies:\n"
        "  NQL form begins with: %s\n"
        "  SQL form begins with: %s"
        % (head, ", ".join(NQL_HEADS), ", ".join(SQL_HEADS))
    )


def execute(db, statement: str):
    """Run a neSQL statement against a Python ``NEDB``, routing on the head.

    The SQL half goes through :mod:`nedb.sql`, which is a TRANSLATOR and has a
    smaller surface than the Rust evaluator — no joins, no subqueries, no
    multiple aggregates. That difference is real and is not hidden: a statement
    the translator cannot express raises rather than returning a partial
    answer, so a caller never receives rows from a different query than the one
    they wrote.
    """
    dialect = route(statement)
    if dialect == "nql":
        return db.query(statement)
    from . import sql as _sql
    return _sql.sql_exec(db, statement)

File: rust/nedb-v2/src/nesql.rs
// SPDX-FileCopyrightText: 2026 INTERCHAINED LLC
// SPDX-License-Identifier: BUSL-1.1
// NEDB · © 2026 INTERCHAINED LLC × Eth-Interchained × Vex (Claude Opus 5)

//! **neSQL** — the whole language: NQL *and* PostgreSQL SQL.
//!
//! Not a third dialect. neSQL is the name for the pair, and this module is the
//! one place that decides which half a statement is written in.
//!
//! # Why the router lives in the engine
//!
//! `nesql-cli` had this logic first, and it was correct there. Putting a
//! second copy in the HTTP server would have been the same mistake this whole
//! effort exists to undo: two implementations of one decision, drifting until
//! `POST /query` and `nesql query` disagree about what a statement means —
//! and disagreeing about the MEANING of a statement is worse than disagreeing
//! about its result, because nothing looks broken.
//!
//! So it moved down here, where both the daemon and the CLI can reach it, and
//! the CLI re-exports it rather than keeping its own.
//!
//! # Routing is structural, so nothing is guessed
//!
//! ```text
//! NQL  begins with FROM
//! SQL  begins with SELECT INSERT UPDATE DELETE EXPLAIN WITH SHOW SET
//!                  VALUES TABLE BEGIN COMMIT ROLLBACK
//! ```
//!
//! PostgreSQL has no statement form that begins with `FROM`, so the leading
//! keyword PARTITIONS the two vocabularies rather than hinting at them. That
//! is what makes this a decision rather than a heuristic, and it is why a
//! first word in neither is REFUSED naming both — never handed to whichever
//! parser seems likelier, because "seems likelier" is the guess the rule
//! forbids.

/// Which half of neSQL a statement is written in — inherited PostgreSQL, or NQL.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Dialect {
    Nql,
    Sql,
}

impl Dialect {
    pub fn name(self) -> &'static str {
        match self {
            Dialect::Nql => "nql",
            Dialect::Sql => "sql",
        }
    }
}

/// Statement-initial keywords, per dialect. Disjoint by construction; the test
/// `the_two_vocabularies_do_not_overlap` holds them that way.
pub const NQL_HEADS: &[&str] = &["FROM"];
pub const SQL_HEADS: &[&str] = &[
    "SELECT", "INSERT", "UPDATE", "DELETE", "EXPLAIN", "WITH", "SHOW", "SET",
    "VALUES", "TABLE", "BEGIN", "COMMIT", "ROLLBACK",
];

fn first_word(s: &str) -> Option<String> {
    s.split_whitespace()
        .next()
        // A statement may open with a parenthesis — `(SELECT …) UNION …`.
        .map(|w| w.trim_start_matches('(').trim_end_matches(';').to_uppercase())
        .filter(|w| !w.is_empty())
}

/// Decide which half a statement is written in, or refuse.
pub fn route(q: &str) -> Result<Dialect, String> {
    let Some(head) = first_word(q) else {
        return Err("the statement is empty".to_string());
    };
    if NQL_HEADS.contains(&head.as_str()) {
        return Ok(Dialect::Nql);
    }
    if SQL_HEADS.contains(&head.as_str()) {
        return Ok(Dialect::Sql);
    }
    Err(format!(
        "{:?} does not begin a neSQL statement.\n\
         neSQL is PostgreSQL's SQL plus NEDB's own clauses, so a statement starts\n\
         in one of these two vocabularies:\n  \
         NQL form begins with: {}\n  \
         SQL form begins with: {}",
        head,
        NQL_HEADS.join(", "),
        SQL_HEADS.join(", "),
    ))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn the_two_vocabularies_do_not_overlap() {
        // The whole no-guessing argument rests on this. If a word ever appears
        // in both lists, routing becomes a coin flip and neSQL starts lying
        // about being total.
        for n in NQL_HEADS {
            assert!(
                !SQL_HEADS.contains(n),
                "{:?} begins a statement in both dialects — routing is no longer structural",
                n
            );
        }
    }

    #[test]
    fn each_dialect_routes_to_itself() {
        assert_eq!(route("FROM orders").unwrap(), Dialect::Nql);
        assert_eq!(route("from orders WHERE x = 1").unwrap(), Dialect::Nql);
        assert_eq!(route("SELECT * FROM orders").unwrap(), Dialect::Sql);
        assert_eq!(route("  explain select 1").unwrap(), Dialect::Sql);
        assert_eq!(route("(SELECT 1) UNION (SELECT 2)").unwrap(), Dialect::Sql);
        assert_eq!(route("INSERT INTO o VALUES (1)").unwrap(), Dialect::Sql);
    }

    #[test]
    fn a_word_in_neither_vocabulary_is_refused_naming_both() {
        let e = route("GRANT ALL ON orders").unwrap_err();
        assert!(e.contains("GRANT"), "{}", e);
        assert!(e.contains("FROM"), "the refusal must name the NQL vocabulary: {}", e);
        assert!(e.contains("SELECT"), "and the SQL one: {}", e);
    }

    #[test]
    fn an_empty_statement_is_refused_rather_than_routed() {
        assert!(route("").is_err());
        assert!(route("   \n ").is_err());
    }

    /// The reason this module is in the engine rather than in the CLI.
    #[test]
    fn a_sql_statement_is_never_handed_to_the_nql_parser() {
        // `POST /query {"nql": "SELECT ..."}` is the case that motivated
        // this: the field is called `nql` for compatibility, and its contents
        // are no longer required to be NQL.
        assert_eq!(route("SELECT who FROM orders").unwrap(), Dialect::Sql);
        assert_eq!(route("FROM orders").unwrap(), Dialect::Nql);
    }
}
